marimbabot_audio/src/audio_from_lilypond.py

```diff
@@ -76,19 +89,59 @@ def callback_vision_results(data: String, audio_publisher):
     # publish audio file to audio node (sound_play package)
     audio_publisher.publish(sound_request)
 
-# create audio from lilypond publisher
-# send audio to audio node (sound_play package)
-def listener():
+    return audio_filename
+
+"""
+Action server for audio generation from lilypond string
+"""
+class AudioFromLilypondActionServer(object):
+    # create messages that are used to publish feedback/result
+    _feedback = LilypondAudioFeedback()
+    _result = LilypondAudioResult()
+
+    def __init__(self, name):
+        # publish to audio node (sound_play package)
+        self.pub = rospy.Publisher('robotsound', SoundRequest, queue_size=50)
+
+        self._action_name = name
+        self._as = actionlib.SimpleActionServer(self._action_name, LilypondAudioAction, execute_cb=self.execute_cb, auto_start = False)
+        self._as.start()
+      
+    def execute_cb(self, goal):
+        # helper variables
+        success = True
+        
+        # append the seeds for the fibonacci sequence
+        self._feedback.in_progress = True
+        
+        # publish info to the console for the user
+        rospy.logdebug("Starting " + self._action_name + " with order " + str(goal.lilypond_string)) 
+        
+        # start executing the action
+        audio_filename = ""
+        try:
+            print("before filename")
+            audio_filename = callback_lilypond_sentence(goal.lilypond_string, self.pub)
+            print(audio_filename)
+            durations_seconds = get_audio_length(audio_filename)
+            time.sleep(durations_seconds)
+            self._feedback.in_progress = False
+        except:
+            # if an error happens during audio generation, set success to false
+            success = False
+            rospy.logerr("Error during audio generation inside audio_from_lilypond action server")
+
+        self._result.success = success
+        if success:
```

Comment:
Add an else case for failures. And you might want to set `in_progress` to `false`

Reply:
This might solve @scaredycode 's issue

Reply:
done

Reply:
I added in_progress and an else case.
I could not replicate the problem of @scaredycode   where the action server cannot be called multiple times / is busy after calling it again.